shortfin/tests/apps/sd/serving_test.py

```diff
@@ -0,0 +1,186 @@
+
+@pytest.mark.system("amdgpu")
+def test_sd_server(sd_server_fpd1):
+    imgs, status_code = send_json_file(sd_server_fpd1.url)
+    assert len(imgs) == 1
+    ref_path = os.path.dirname(__file__)
+    ref_path = os.path.join(ref_path, "refs/snow_cat.png")
+    if not os.path.exists(ref_path):
+        assert 0, f"Reference image does not exist at {ref_path}"
+    ref_image = Image.open(ref_path)
+    ref_image_numpy = numpy.array(ref_image)
+    gen_image_numpy = numpy.array(imgs[0])
+    assert (ref_image_numpy == gen_image_numpy).all()
+    assert status_code == 200
+
+
+class ServerRunner:
+    def __init__(self, args):
+        port = str(find_free_port())
+        self.url = "http://0.0.0.0:" + port
+        env = os.environ.copy()
+        env["PYTHONUNBUFFERED"] = "1"
+        self.process = subprocess.Popen(
+            [
+                *args,
+                "--port=" + port,
+                "--device=amdgpu",
+            ],
+            env=env,
+            # TODO: Have a more robust way of forking a subprocess.
+            stdout=sys.stdout,
+            stderr=sys.stderr,
+        )
+        self._wait_for_ready()
+
+    def _wait_for_ready(self):
+        start = time.time()
+        while True:
+            time.sleep(2)
+            try:
+                if requests.get(f"{self.url}/health").status_code == 200:
+                    return
+            except Exception as e:
+                if self.process.errors is not None:
+                    raise RuntimeError("API server process terminated") from e
+            time.sleep(1.0)
+            if (time.time() - start) > 60:
+                raise RuntimeError("Timeout waiting for server start")
+
+    def __del__(self):
+        try:
+            process = self.process
+        except AttributeError:
+            pass
+        else:
+            process.terminate()
+            process.wait()
+
+
+def send_json_file(url="http://0.0.0.0:8000", num_copies=1):
+    # Read the JSON file
+    data = copy.deepcopy(sample_request)
+    imgs: list[Image.Image] = []
+    # Send the data to the /generate endpoint
+    data["prompt"] = (
+        [data["prompt"]]
+        if isinstance(data["prompt"], str)
+        else data["prompt"] * num_copies
+    )
+    try:
+        response = requests.post(url + "/generate", json=data)
+        response.raise_for_status()  # Raise an error for bad responses
+        response_body = response.json()
+
+        for idx, each_png in enumerate(response_body["images"]):
+            if not isinstance(each_png, str):
+                raise ValueError(
+                    f"Expected string-encoded png at index {idx}, found {each_png}"
+                )
+
+            each_image = image_from(Base64CharacterEncodedByteSequence(each_png))
+            imgs.append(each_image)
+
+    except requests.exceptions.RequestException as e:
+        print(f"Error sending the request: {e}")
+
+    return imgs, response.status_code
+
+
+def find_free_port():
```

Comment:
I've seen similar code in other parts of shortfin. Can we use a common helper function instead of reimplementing? (We should have multiple tests using servers like this, so such tests should have a robust library of utils to call on)